srai/regionizers/h3_regionizer.py

```diff
@@ -0,0 +1,194 @@
+"""
+H3 regionizer.
+
+This module exposes Uber's H3 Hexagonal Hierarchical Geospatial Indexing System [1] as a regionizer.
+
+Note:
+    The default API [2] was chosen (basic_str) to ease the implementation.
+    It may be beneficial to try the NumPy API for computationally-heavy work.
+
+References:
+    [1] https://uber.github.io/h3-py/
+    [2] https://uber.github.io/h3-py/api_comparison
+
+"""
+
+from typing import List
+
+import geopandas as gpd
+import h3
+import pandas as pd
+from functional import seq
+from shapely import geometry
+
+
+class H3Regionizer:
+    """
+    H3 Regionizer.
+
+    H3 Regionizer allows the given geometries to be divided
+    into H3 cells - hexagons with pentagons as a very rare exception
+
+    """
+
+    def __init__(self, resolution: int, buffer: bool = True) -> None:
+        """
+        Init H3Regionizer.
+
+        Args:
+            resolution (int): Resolution of the cells. See [1] for a full comparison.
+            buffer (bool, optional): Whether to fully cover the geometries with
+                H3 Cells (visible on the borders).
+                Turn off for large geometries, as it's computationally expensive.
+                Defaults to True.
+
+        Raises:
+            ValueError: If resolution is not between 0 and 15.
+
+        References:
+            [1] https://h3geo.org/docs/core-library/restable/
+
+        """
+        if not (0 <= resolution <= 15):
+            raise ValueError(f"Resolution {resolution} is not between 0 and 15.")
+
+        self.resolution = resolution
+        self.buffer = buffer
+
+    def transform(self, gdf: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
+        """
+        Regionize a given GeoDataFrame.
+
+        Transforms given geometries into H3 cells of given resolution
+        and optionally applies buffering.
+
+        Args:
+            gdf (gpd.GeoDataFrame): (Multi)Polygons to be regionized.
+
+        Returns:
+            gpd.GeoDataFrame: H3 cells.
+
+        Raises:
+            ValueError: If provided GeoDataFrame has no crs defined.
+
+        """
+        gdf_wgs84 = gdf.to_crs(epsg=4326)
+        # transform multipolygons to multiple polygons
+        gdf_exploded = gdf_wgs84.explode(index_parts=True).reset_index(drop=True)
+
+        h3_indexes = (
+            seq(gdf_exploded["geometry"])
+            .map(self._polygon_shapely_to_h3)
+            .flat_map(lambda polygon: h3.polygon_to_cells(polygon, self.resolution))
+            .to_list()
+        )
+
+        gdf_h3 = self._gdf_from_h3_indexes(h3_indexes)
+        if self.buffer:
+            return self._buffer(gdf_exploded, gdf_h3)
+
+        return gdf_h3.to_crs(gdf.crs)
```

Comment:
Is there a reason for this to be so complicated? IIRC, my solution to this was just:
1. buffer - 2 * edge length or something,
2. polyfill,
3. sjoin with original polygon to remove non-overlapping hexes